### PR TITLE
Feature/wcwpp 46 add fulfillment request status settings

### PR DESCRIPTION
**Pull request by lucapircher**
## TLDR;

- adds the fulfilment settings to the backend
- converts the singleton to use lazy initialization to make testing easier
  During initialization the plugin loads the available settigns using the WooCommerce functions. Since WooCommerce is not available for unit tests we have to mock them. With lazy initialization we can easily provide a mock implementation for the `OrderStateRepository`. 
- adds validation for the fulfilment settings which are triggered when the wordpress backend is loaded
- refactors the plugin to use the singleton base class
- changes the initialization to rely on the autoloader instead of manually loading the plugin class

## Additional Info

If you have an idea on how to propperly test the validation of the current settings for the fulfilment settings, tell me please. I stopped atempting it before wasting to much time.

### CODE REVIEW

File: includes/FulfilmentOptionsTab.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<?php
+/**
+ * Fulfilment Options Tab
+ *
+ * @package GbWeissOptions
+ */
+
+namespace GbWeiss\includes;
```
Comment (Martin-Welte) [inline includes/FulfilmentOptionsTab.php]: protect class from direct access

File: includes/GbWeiss.php [hunk 2]
```diff
@@ -9,12 +9,12 @@
...
 namespace GbWeiss\includes;
 
-defined('ABSPATH') || exit;
+use GbWeiss\includes\OrderStateRepository;
 
 /**
  * Main GbWeiss class
  */
-final class GbWeiss
+final class GbWeiss extends Singleton
```
Comment (Martin-Welte) [inline includes/GbWeiss.php:17]: man i could have thought about this. Nice 👍

File: includes/GbWeiss.php [hunk 3]
```diff
@@ -80,6 +76,11 @@ public function initOptionPage(): void
             ->addOption(new Option('Client Secret', 'client_secret', __('Client Secret', self::$languageDomain), 'account', 'string'));
 
         $optionsPage->addTab($accountTab);
+        $orderStatuses = $this->orderStateRepository->getAllOrderStates();
```
Comment (philipp-hemmer) [inline includes/GbWeiss.php:79]: i'd write orderStates

File: includes/GbWeiss.php [hunk 4]
```diff
@@ -118,6 +119,51 @@ public static function checkPluginCompatabilityAndPrintErrorMessages(): bool
...
+     * @param string $optionName The name of the WordPress option.
+     * @param string $displayName The name of the setting to be shown in error messages.
+     * @return void
+     */
+    private function validateFulfilmentSetting(string $optionName, string $displayName): void
+    {
+        $optionValue = \get_option($optionName);
+
+        if (!$optionValue) {
+            $this->showWordpressAdminErrorMessage("The Gebrüder Weiss WooCommerce Plugin settings are missing a value for " . $displayName . ".");
```
Comment (Martin-Welte) [inline includes/GbWeiss.php]: The plugin is only single language, but i would prepare multilanguage already.

File: includes/GbWeiss.php [hunk 5]
```diff
@@ -118,6 +119,51 @@ public static function checkPluginCompatabilityAndPrintErrorMessages(): bool
...
+            return;
+        }
+
+        if (!$this->checkIfWooCommerceOrderStateExists($optionValue)) {
+            $this->showWordpressAdminErrorMessage("The selected order state for " . $displayName . " in the options for the Gebrüder Weiss WooCommerce Plugin does not exist in WooCommerce.");
```
Comment (Martin-Welte) [inline includes/GbWeiss.php]: also here

File: includes/GbWeiss.php [hunk 6]
```diff
@@ -80,6 +76,11 @@ public function initOptionPage(): void
             ->addOption(new Option('Client Secret', 'client_secret', __('Client Secret', self::$languageDomain), 'account', 'string'));
 
         $optionsPage->addTab($accountTab);
+        $orderStatuses = $this->orderStateRepository->getAllOrderStates();
+
+        $fulfilmentTab = new FulfilmentOptionsTab($orderStatuses);
```
Comment (philipp-hemmer) [inline includes/GbWeiss.php]: fulfillment :)

Comment (philipp-hemmer) [reply includes/GbWeiss.php]: find and replace it, since the typie is also in the name "validateFulfilmentSetting"

File: includes/GbWeiss.php (hunk 4 above)
Comment (philipp-hemmer) [inline includes/GbWeiss.php]: i'm not sure, but i think i saw translations in martins PR yesterday

**Review by Martin-Welte — CHANGES_REQUESTED** [review]

**Review by philipp-hemmer — CHANGES_REQUESTED** [review]
some typies

File: includes/GbWeiss.php (hunk 3 above)
Comment (lucapircher) [reply includes/GbWeiss.php:79]: To my knowledge the singular entity name is used when following the repository pattern.

Comment (philipp-hemmer) [reply includes/GbWeiss.php:79]: but the term "statuses" does not exist in the language. could be "status" or "stati" ... but this is "haarspalterei" ;)

Comment (lucapircher) [reply includes/GbWeiss.php:79]: The name actually comes from WooCommerce, but I'll change it.

**Review by Martin-Welte — APPROVED** [review]

**Review by philipp-hemmer — APPROVED** [review]
